stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -1,56 +1,560 @@
+          Bd_(ChainableStack::memalloc_.alloc_array<double>(B_size_)),
+          variRefB_(ChainableStack::memalloc_.alloc_array<vari*>(B_size_)),
+          variRefAB_(ChainableStack::memalloc_.alloc_array<vari*>(A_rows_
+                                                                  * B_cols_)) {
+            using Eigen::MatrixXd;
+            using Eigen::Map;
+            for (size_type i = 0; i < A.size(); ++i)
+              Ad_[i] = A.coeffRef(i);
+            for (size_type i = 0; i < B.size(); ++i) {
+              variRefB_[i] = B.coeffRef(i).vi_;
+              Bd_[i] = B.coeffRef(i).val();
+            }
+            MatrixXd AB
+              = Map<MatrixXd>(Ad_, A_rows_, A_cols_)
+              * Map<MatrixXd>(Bd_, A_cols_, B_cols_);
+            for (size_type i = 0; i < AB.size(); ++i)
+              variRefAB_[i] = new vari(AB.coeffRef(i), false);
+          }
+
+        virtual void chain() {
+          using Eigen::MatrixXd;
+          using Eigen::Map;
+          MatrixXd adjAB(A_rows_, B_cols_);
+          MatrixXd adjB(A_cols_, B_cols_);
+
+          for (size_type i = 0; i < adjAB.size(); ++i)
+            adjAB(i) = variRefAB_[i]->adj_;
+          adjB = Map<MatrixXd>(Ad_, A_rows_, A_cols_).transpose()
+            * adjAB;
+          for (size_type i = 0; i < B_size_; ++i)
+            variRefB_[i]->adj_ += adjB(i);
+        }
+    };
+
+    /**
+     * This is a subclass of the vari class for matrix 
+     * multiplication A * B where A is a double
+     * row vector of length M and B is a vector of
+     * length M.
+     *
+     * The class stores the structure of each matrix,
+     * the double values of A and B, and pointers to
+     * the varis for A and B if A or B is a var. It
+     * also instantiates and stores pointers to
+     * varis for all elements of A * B.
+     *
+     * @tparam CA Columns for matrix A, Rows for matrix B
+     * @tparam TB Scalar type for matrix B
+     */
+    template <int CA, typename TB>
+    class multiply_mat_vari<double, 1, CA, TB, 1> : public vari {
+      public:
+        int A_cols_;
+        double* Ad_;
+        double* Bd_;
+        vari** variRefB_;
+        vari* variRefAB_;
+
+        /** 
+         * Constructor for multiply_matrix_vari.
+         *
+         * All memory allocated in
+         * ChainableStack's stack_alloc arena.
+         *
+         * It is critical for the efficiency of this object
+         * that the constructor create new varis that aren't
+         * popped onto the var_stack_, but rather are 
+         * popped onto the var_nochain_stack_. This is
+         * controlled to the second argument to 
+         * vari's constructor.
+         *
+         * @param A row vector
+         * @param B vector
+         */
+        multiply_mat_vari(const Eigen::Matrix<double, 1, CA>& A,
+                          const Eigen::Matrix<TB, CA, 1>& B)
+          : vari(0.0),
+          A_cols_(A.cols()),
```

Comment:
why not call this something like `size_`? It gets confusing in line 329 where you use `A_cols_` for the size of `B`.